From a1ff11f2abe51525a8f2db7e9bbf984a99fc5002 Mon Sep 17 00:00:00 2001
From: Luca Bacci <luca.bacci982@gmail.com>
Date: Sun, 4 Sep 2022 15:16:21 +0200
Subject: [PATCH 13/14] Check for scaled_font::cache_frozen in
 glyph_page_can_remove()

With modifications kindly suggested by Uli Schlachter

Fixes https://gitlab.freedesktop.org/cairo/cairo/-/issues/587

(cherry picked from commit 166d718099fc8898895d4ad2839d5fb9865a87a5)
---
 src/cairo-scaled-font.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/cairo-scaled-font.c b/src/cairo-scaled-font.c
index 42fb55915..973d5469a 100755
--- a/src/cairo-scaled-font.c
+++ b/src/cairo-scaled-font.c
@@ -2898,7 +2898,16 @@ _cairo_scaled_glyph_page_can_remove (const void *closure)
     cairo_scaled_font_t *scaled_font;
 
     scaled_font = page->scaled_font;
-    return CAIRO_MUTEX_TRY_LOCK (scaled_font->mutex);
+
+    if (!CAIRO_MUTEX_TRY_LOCK (scaled_font->mutex))
+       return FALSE;
+
+    if (scaled_font->cache_frozen != 0) {
+       CAIRO_MUTEX_UNLOCK (scaled_font->mutex);
+       return FALSE;
+    }
+
+    return TRUE;
 }
 
 static cairo_status_t
-- 
2.37.0 (Apple Git-136)

